### PR TITLE
chore: add errors to swapper

### LINKED ISSUE #540: Integrate `Errors` package in Swapper - Lib
### Overview

- Currently the error handling in Swapper is tightly coupled with Zrx, we need to use a general package to avoid error handling coupling.

- Call `CreateErrorClass` ("SwapperError") which will return the error class.
- In the caller, you can use it by `throw SwapperError`
- [Errors Package](https://github.com/shapeshift/lib/tree/main/packages/errors)

### References and additional details

- Refer to `Errors` Package in `lib`

### Acceptance Criteria

- Create an Error handling strategy that is not coupled with any one swapper implementation.
- The Errors package needs to be a `peerDependency`

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

Day - 2 days

### PR DESCRIPTION
**Pull request by DaoDev44**
- Add new error handling to swappers
- Remove unnecessary guard checks
- dont rely on symbol or chain, use assetId or chainId instead
closes #540

### CODE REVIEW

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts [hunk 1]
```diff
@@ -1,48 +1,58 @@
...
-    gasPrice: quote?.feeData?.chainSpecific?.gasPrice
+  try {
+    if (sellAsset.chainId !== 'eip155:1') {
+      throw new ZrxSwapError('[ZrxApprovalNeeded] - sellAsset chainId is not supported', {
+        code: SwapErrorTypes.APPROVAL_NEEDED,
+        details: { chainId: sellAsset.chainId }
+      })
+    }
+
+    if (sellAsset.symbol === 'ETH') {
```
Comment (elmutt) [inline packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:24]: compare caip19

Comment (0xdef1cafe) [reply packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:24]: better still, grab the supported chains from chain adapters, get the native token from the chain id, and do `![feeAssets].includes(sellAsset.id)`

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts [hunk 2]
```diff
@@ -1,48 +1,58 @@
...
-    ownerAddress: receiveAddress
-  })
-  const allowanceOnChain = bnOrZero(allowanceResult)
-
-  return {
-    approvalNeeded: allowanceOnChain.lte(bnOrZero(quote.sellAmount)),
-    gas: APPROVAL_GAS_LIMIT,
-    gasPrice: quote?.feeData?.chainSpecific?.gasPrice
+  try {
+    if (sellAsset.chainId !== 'eip155:1') {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:17]: the swapper should have some `public static supportedChainIds = AssetId[]`

then we can 

`if (!this.supportedChainIds.includes(sellAsset.chainId))`

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts [hunk 3]
```diff
@@ -1,48 +1,58 @@
...
+        code: SwapErrorTypes.APPROVAL_NEEDED,
+        details: { chainId: sellAsset.chainId }
+      })
+    }
+
+    if (sellAsset.symbol === 'ETH') {
+      return { approvalNeeded: false }
+    }
+
+    const accountNumber = quote.sellAssetAccountId ? Number(quote.sellAssetAccountId) : 0
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:28]:
```suggestion
    const accountNumber = bnOrZero(quote.sellAssetAccountId).toNumber()
```

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts [hunk 4]
```diff
@@ -1,48 +1,58 @@
...
+      tokenId: quote.sellAsset.tokenId,
+      spenderAddress: quote.allowanceContract,
+      ownerAddress: receiveAddress
+    })
+    const allowanceOnChain = bnOrZero(allowanceResult)
+
+    return {
+      approvalNeeded: allowanceOnChain.lte(bnOrZero(quote.sellAmount)),
+      gas: APPROVAL_GAS_LIMIT,
+      gasPrice: quote?.feeData?.chainSpecific?.gasPrice
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:50]: do some logic to check for existence of `gasPrice` and throw if we don't have it

File: packages/swapper/src/swappers/zrx/ZrxApproveInfinite/ZrxApproveInfinite.ts [hunk 5]
```diff
@@ -1,26 +1,33 @@
...
-    wallet,
-    adapterManager,
-    erc20Abi,
-    web3
-  })
+  try {
+    const allowanceGrantRequired = await grantAllowance({
+      quote: {
+        ...quote,
+        allowanceContract: quote.allowanceContract,
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/ZrxApproveInfinite/ZrxApproveInfinite.ts:17]:
```suggestion
```
gets spread above

File: packages/swapper/src/swappers/zrx/getZrxMinMax/getZrxMinMax.ts [hunk 6]
```diff
@@ -1,24 +1,29 @@
...
+    }
 
-  const usdRate = await getUsdRate({
-    symbol: sellAsset.symbol,
-    tokenId: sellAsset.tokenId
-  })
+    const usdRate = await getUsdRate({
+      symbol: sellAsset.symbol,
+      tokenId: sellAsset.tokenId
+    })
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/getZrxMinMax/getZrxMinMax.ts:18]:
```suggestion
    const usdRate = await getUsdRate({ ...sellAsset })
```

File: packages/swapper/src/swappers/zrx/getZrxMinMax/getZrxMinMax.ts [hunk 7]
```diff
@@ -1,24 +1,29 @@
...
-  const minimum = bn(1).dividedBy(bnOrZero(usdRate)).toString()
+    const minimum = bn(1).dividedBy(bnOrZero(usdRate)).toString()
 
-  return {
-    minimum, // $1 worth of the sell token.
-    maximum: MAX_ZRX_TRADE // Arbitrarily large value. 10e+28 here.
+    return {
+      minimum, // $1 worth of the sell token.
+      maximum: MAX_ZRX_TRADE // Arbitrarily large value. 10e+28 here.
+    }
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/getZrxMinMax/getZrxMinMax.ts:25]:
```suggestion
    const minimum = bn(1).dividedBy(bnOrZero(usdRate)).toString() // $1 worth of the sell token.
    const maximum = MAX_ZRX_TRADE // Arbitrarily large value. 10e+28 here.
    return { minimum, maximum }
```

File: packages/swapper/src/swappers/zrx/getZrxTradeQuote/getZrxTradeQuote.ts [hunk 8]
```diff
@@ -1,25 +1,22 @@
...
 
 export async function getZrxTradeQuote(input: GetTradeQuoteInput): Promise<TradeQuote<'eip155:1'>> {
   const { sellAsset, buyAsset, sellAmount, sellAssetAccountId } = input
-  if (!buyAsset) {
-    throw new ZrxError('getQuote - Missing buyAsset')
-  }
-  if (!sellAsset) {
-    throw new ZrxError('getQuote - Missing sellAsset')
-  }
   if (buyAsset.chainId !== 'eip155:1' || sellAsset.chainId !== 'eip155:1') {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/getZrxTradeQuote/getZrxTradeQuote.ts:15]: use every on array of sellAsset and buyAsset vs  suggested`this.supportedChainIds` to determine this

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 9]
```diff
@@ -69,56 +69,76 @@ export const getAllowanceRequired = async ({
   web3,
   erc20AllowanceAbi
 }: GetAllowanceRequiredArgs): Promise<BigNumber> => {
-  if (sellAsset.symbol === 'ETH') {
-    return bn(0)
-  }
+  try {
+    if (sellAsset.assetId === 'eip155:1/slip44:60') {
+      return bn(0)
+    }
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:75]: same deal supportedChainIds or similar

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 10]
```diff
@@ -69,56 +69,76 @@ export const getAllowanceRequired = async ({
...
+    const allowanceOnChain = await getERC20Allowance({
+      web3,
+      erc20AllowanceAbi,
+      ownerAddress,
+      spenderAddress,
+      tokenId
+    })
+    if (allowanceOnChain === '0') {
+      return bnOrZero(sellAmount)
+    }
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:96]: one line

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 11]
```diff
@@ -69,56 +69,76 @@ export const getAllowanceRequired = async ({
...
+    if (sellAsset.assetId === 'eip155:1/slip44:60') {
+      return bn(0)
+    }
 
-  const ownerAddress = receiveAddress
-  const spenderAddress = allowanceContract
-  const tokenId = sellAsset.tokenId
+    const ownerAddress = receiveAddress
+    const spenderAddress = allowanceContract
+    const tokenId = sellAsset.tokenId
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:79]: destructure

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 12]
```diff
@@ -69,56 +69,76 @@ export const getAllowanceRequired = async ({
...
-  if (symbol === 'USDC') return '1' // Will break if comparing against usdc
-  const rateResponse: AxiosResponse<ZrxPriceResponse> = await zrxService.get<ZrxPriceResponse>(
-    '/swap/v1/price',
-    {
-      params: {
-        buyToken: 'USDC',
-        buyAmount: '1000000000', // rate is imprecise for low $ values, hence asking for $1000
-        sellToken: tokenId || symbol
+  try {
+    if (symbol === 'USDC') return '1' // Will break if comparing against usdc
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:116]: at least use assetId, we should *only* ever use symbol when talking to zrx

better option - use contract addresses if zrx supports it

best option - if zrx has an endpoint that gives us all the available symbols, add a caip adapter for zrx, e.g.

`type BuyToken = getZrxId(assetId: AssetId) => string`

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 13]
```diff
@@ -69,56 +69,76 @@ export const getAllowanceRequired = async ({
...
+    const rateResponse: AxiosResponse<ZrxPriceResponse> = await zrxService.get<ZrxPriceResponse>(
+      '/swap/v1/price',
+      {
+        params: {
+          buyToken: 'USDC',
+          buyAmount: '1000000000', // rate is imprecise for low $ values, hence asking for $1000
+          sellToken: tokenId || symbol
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:123]: looks like it does take contract addresses

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 14]
```diff
@@ -128,23 +148,25 @@ export const grantAllowance = async ({
...
+        code: SwapErrorTypes.GRANT_ALLOWANCE
+      })
+    }
 
-  const adapter = await adapterManager.byChainId('eip155:1')
-  const erc20Contract = new web3.eth.Contract(erc20Abi, quote.sellAsset.tokenId)
-  const approveTx = erc20Contract.methods
-    .approve(quote.allowanceContract, quote.sellAmount)
-    .encodeABI()
+    const adapter = await adapterManager.byChainId('eip155:1')
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:158]: supported chain ids blah

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 15]
```diff
@@ -128,23 +148,25 @@ export const grantAllowance = async ({
...
+    const approveTx = erc20Contract.methods
+      .approve(quote.allowanceContract, quote.sellAmount)
+      .encodeABI()
 
-  const bip44Params = adapter.buildBIP44Params({
-    accountNumber: Number(quote.sellAssetAccountId) || 0
-  })
+    const bip44Params = adapter.buildBIP44Params({
+      accountNumber: Number(quote.sellAssetAccountId) || 0
+    })
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:166]:
```suggestion
    const accountNumber = bnOrZero(quote.sellAssetAccountId).toNumber()
    const bip44Params = adapter.buildBIP44Params({ accountNumber })
```

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 16]
```diff
@@ -128,23 +148,25 @@ export const grantAllowance = async ({
...
 
-  let grantAllowanceTxToSign, signedTx, broadcastedTxId
+    let signedTx, broadcastedTxId
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:168]: IIFE pattern plz

File: packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts [hunk 17]
```diff
@@ -1,40 +1,23 @@
...
-  const adapter = adapterManager.byChain(sellAsset.chain)
-  const bip44Params = adapter.buildBIP44Params({
-    accountNumber: Number(trade.sellAssetAccountId)
-  })
-
-  let buildTxResponse, signedTx, txid
   try {
-    buildTxResponse = await adapter.buildSendTransaction({
+    // value is 0 for erc20s
+    const value = sellAsset.symbol === 'ETH' ? trade.sellAmount : '0'
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts:14]: assetId

File: packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts [hunk 18]
```diff
@@ -1,40 +1,23 @@
...
+    const adapter = adapterManager.byChain(sellAsset.chain)
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts:15]:
```suggestion
    const adapter = adapterManager.byChainId(sellAsset.chainId)
```

File: packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts [hunk 19]
```diff
@@ -1,40 +1,23 @@
...
+    const bip44Params = adapter.buildBIP44Params({
+      accountNumber: Number(trade.sellAssetAccountId)
+    })
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts:18]: bnOrZero and one line this

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts (hunk 1 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:24]: not sure this is possible from chain adapters currently

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts (hunk 14 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:158]: https://github.com/shapeshift/lib/issues/646

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts (hunk 9 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:75]: https://github.com/shapeshift/lib/issues/646

File: packages/swapper/src/swappers/zrx/getZrxTradeQuote/getZrxTradeQuote.ts (hunk 8 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/getZrxTradeQuote/getZrxTradeQuote.ts:15]: https://github.com/shapeshift/lib/issues/646

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts (hunk 2 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:17]: https://github.com/shapeshift/lib/issues/646

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts (hunk 11 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:79]: stylistic

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts (hunk 12 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:114]: contract address on asset should be what we what we need

File: packages/swapper/src/api.ts [hunk 20]
```diff
@@ -109,7 +112,20 @@ export type ApprovalNeededInput<C extends SupportedChainIds> = {
   wallet: HDWallet
 }
 
-export class SwapError extends Error {}
+// Swap Errors
+export enum SwapErrorTypes {
+  SIGN_AND_BROADCAST_FAILED = 'SIGN_AND_BROADCAST_FAILED',
+  UNSUPPORTED_PAIR = 'UNSUPPORTED_PAIR',
+  MIN_MAX_ERROR = 'MIN_MAX_ERROR ',
```
Comment (0xdef1cafe) [inline packages/swapper/src/api.ts:119]:
```suggestion
  MIN_MAX_ERROR = 'MIN_MAX_ERROR',
```

File: packages/swapper/src/manager/SwapperManager.test.ts [hunk 21]
```diff
@@ -64,15 +64,15 @@ describe('SwapperManager', () => {
...
+        '[getSwapper] - swapperType doesnt exist'
       )
     })
 
     it('should throw an error if an invalid Swapper instance is passed', () => {
       const manager = new SwapperManager()
       // @ts-ignore
       expect(() => manager.addSwapper(SwapperType.Thorchain, {})).toThrow(
-        'SwapperError:validateSwapper - invalid swapper instance'
+        '[getSwapper] - invalid swapper instanc'
```
Comment (0xdef1cafe) [inline packages/swapper/src/manager/SwapperManager.test.ts:75]: missing `e`

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts [hunk 22]
```diff
@@ -13,36 +13,54 @@ export async function ZrxApprovalNeeded(
...
+        code: SwapErrorTypes.APPROVAL_NEEDED,
+        details: { chainId: sellAsset.chainId }
+      })
+    }
 
-  if (sellAsset.symbol === 'ETH') {
-    return { approvalNeeded: false }
-  }
+    // No approval needed for trading eth itself
+    if (sellAsset.assetId === 'eip155:1/slip44:60') {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:25]: 1. get the fee asset id
2. if the asset id === fee asset id return false

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts [hunk 23]
```diff
@@ -13,36 +13,54 @@ export async function ZrxApprovalNeeded(
...
-  const bip44Params = adapter.buildBIP44Params({ accountNumber })
-  const receiveAddress = await adapter.getAddress({ wallet, bip44Params })
+    const adapter = await adapterManager.byChainId(sellAsset.chainId)
+    const bip44Params = adapter.buildBIP44Params({ accountNumber })
+    const receiveAddress = await adapter.getAddress({ wallet, bip44Params })
 
-  if (!quote.sellAsset.tokenId || !quote.allowanceContract) {
-    throw new SwapError('ZrxApprovalNeeded - tokenId and allowanceTarget are required')
-  }
+    if (!quote.sellAsset.tokenId || !quote.allowanceContract) {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:35]: i don't want to see `tokenId` in new code - this should actually be removed from the asset type. all the information is encoded in the assetId - do this instead

```
const { assetReference } = fromAssetId(quote.sellAsset.assetId) // per CAIP19, assetReference *is* the tokenId
```

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 24]
```diff
@@ -69,56 +68,75 @@ export const getAllowanceRequired = async ({
...
+    if (sellAsset.assetId === 'eip155:1/slip44:60') {
+      return bn(0)
+    }
 
-  const ownerAddress = receiveAddress
-  const spenderAddress = allowanceContract
-  const tokenId = sellAsset.tokenId
+    const ownerAddress = receiveAddress
+    const spenderAddress = allowanceContract
+    const tokenId = sellAsset.tokenId
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:78]: similarly, don't use this

File: packages/swapper/src/swappers/zrx/zrxBuildTrade/zrxBuildTrade.ts [hunk 25]
```diff
@@ -32,39 +31,20 @@ export async function zrxBuildTrade(
...
-    throw new SwapError(
-      'ZrxSwapper:ZrxBuildTrade One of buyAssetContract or buyAssetSymbol or buyAssetNetwork are required'
-    )
-  }
-  if (!sellToken) {
-    throw new SwapError(
-      'ZrxSwapper:ZrxBuildTrade One of sellAssetContract or sellAssetSymbol or sellAssetNetwork are required'
-    )
-  }
+  const buyToken = buyAsset.tokenId || buyAsset.symbol
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/zrxBuildTrade/zrxBuildTrade.ts:34]: symbol is dangerous - add comment to default to a supported fee asset id for a given swapper in future

File: packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts [hunk 26]
```diff
@@ -1,40 +1,23 @@
...
-  })
-
-  let buildTxResponse, signedTx, txid
   try {
-    buildTxResponse = await adapter.buildSendTransaction({
+    // value is 0 for erc20s
+    const value = sellAsset.assetId === 'eip155:1/slip44:60' ? trade.sellAmount : '0'
+    const adapter = await adapterManager.byChainId(sellAsset.chainId)
+    const bip44Params = adapter.buildBIP44Params({
+      accountNumber: bnOrZero(trade.sellAssetAccountId).toNumber()
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts:17]: separate ticket to rename this field on the trade type from `sellAssetAccountId` -> `sellAssetAccountNumber`

**Review by 0xdef1cafe — COMMENTED** [review]
see comments

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts (hunk 23 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:35]: https://github.com/shapeshift/lib/issues/652

File: packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts (hunk 26 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/zrxExecuteTrade/zrxExecuteTrade.ts:17]: https://github.com/shapeshift/lib/issues/653

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts (hunk 24 above)
Comment (elmutt) [reply packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:78]: https://github.com/shapeshift/lib/issues/653

**Review by elmutt — APPROVED** [review]

File: packages/swapper/src/api.ts [hunk 27]
```diff
@@ -109,7 +112,21 @@ export type ApprovalNeededInput<C extends SupportedChainIds> = {
...
+  MIN_MAX_ERROR = 'MIN_MAX_ERROR',
+  APPROVAL_NEEDED = 'APPROVAL_NEEDED',
+  APPROVE_INFINITE = 'APPROVE_INFINITE',
+  EXECUTE_TRADE = 'EXECUTE_TRADE',
+  ALLOWANCE_REQUIRED = 'ALLOWANCE_REQUIRED',
+  USD_RATE_FAILED = 'USD_RATE_FAILED',
+  GRANT_ALLOWANCE = 'GRANT_ALLOWANCE',
+  BUILD_TRADE = 'BUILD_TRADE',
+  MANAGER = 'MANAGER',
+  TRADE_QUOTE = 'TRADE_QUOTE'
```
Comment (cjthompson) [inline packages/swapper/src/api.ts:128]: sort

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts [hunk 28]
```diff
@@ -13,36 +13,54 @@ export async function ZrxApprovalNeeded(
...
+      throw new SwapError('[ZrxApprovalNeeded] - no gas price with quote', {
+        code: SwapErrorTypes.APPROVAL_NEEDED
+      })
+    return {
+      approvalNeeded: allowanceOnChain.lte(bnOrZero(quote.sellAmount)),
+      gas: APPROVAL_GAS_LIMIT,
+      gasPrice: quote.feeData.chainSpecific?.gasPrice
+    }
+  } catch (e) {
+    throw new SwapError('[ZrxApprovalNeeded]', {
```
Comment (cjthompson) [inline packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:61]:
```
if (e instanceof SwapError) throw e
else { throw new SwapError()... }
```

File: packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts [hunk 29]
```diff
@@ -13,36 +13,54 @@ export async function ZrxApprovalNeeded(
...
-  const allowanceOnChain = bnOrZero(allowanceResult)
-
-  return {
-    approvalNeeded: allowanceOnChain.lte(bnOrZero(quote.sellAmount)),
-    gas: APPROVAL_GAS_LIMIT,
-    gasPrice: quote?.feeData?.chainSpecific?.gasPrice
+    if (!quote.feeData.chainSpecific?.gasPrice)
+      throw new SwapError('[ZrxApprovalNeeded] - no gas price with quote', {
+        code: SwapErrorTypes.APPROVAL_NEEDED
+      })
```
Comment (cjthompson) [inline packages/swapper/src/swappers/zrx/ZrxApprovalNeeded/ZrxApprovalNeeded.ts:54]:
```suggestion
      throw new SwapError('[ZrxApprovalNeeded] - no gas price with quote', {
        code: SwapErrorTypes.APPROVAL_NEEDED,
        details: { fn: 'ZrxApprovalNeeded', feeData: quote.feeData }
      })
```

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 30]
```diff
@@ -69,56 +68,75 @@ export const getAllowanceRequired = async ({
...
+      erc20AllowanceAbi,
+      ownerAddress,
+      spenderAddress,
+      tokenId
+    })
+    if (allowanceOnChain === '0') return bnOrZero(sellAmount)
+    if (!allowanceOnChain) {
+      throw new SwapError(
+        `[getAllowanceRequired] - No allowance data for ${allowanceContract} to ${receiveAddress}`,
+        { code: SwapErrorTypes.ALLOWANCE_REQUIRED }
```
Comment (cjthompson) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:97]: put those into details

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 31]
```diff
@@ -157,43 +174,38 @@ export const grantAllowance = async ({
...
+    if (wallet.supportsOfflineSigning()) {
+      const signedTx = await adapter.signTransaction({ txToSign: grantAllowanceTxToSign, wallet })
 
-    if (!signedTx) {
-      throw new SwapError(`grantAllowance - Signed transaction is required: ${signedTx}`)
-    }
+      if (!signedTx) {
+        throw new SwapError(`[grantAllowance] - Signed transaction is required: ${signedTx}`, {
+          code: SwapErrorTypes.GRANT_ALLOWANCE
+        })
```
Comment (cjthompson) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:187]: put signedtx into details

**Review by cjthompson — APPROVED** [review]

**Review by cjthompson — APPROVED** [review]

**Review by elmutt — APPROVED** [review]

**Pull request merged**